File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include<iostream>
#include<vector>

using namespace std;

int N, M;
vector<int> arr;
bool vis[10];


void DFS(int n) {
	if (n == M) {
		for (int i = 0; i < M; i++) {
			cout << arr[i] << ' ';
		}
		cout<<'\n';
		return;
	}
	else {

		for (int i = 1; i <= N; i++) {
			if (!vis[i]) {
				arr.push_back(i);
				vis[i] = 1;
				DFS(n + 1);
				arr.pop_back();
				vis[i] = 0;
			}
		}
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	cin >> N >> M;
	DFS(0);
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java

import java.io.*;
import java.util.*;

public class Main {

    static int[] arr;
    static boolean[] vis;
    static int N,M;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st =new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[M];
        vis = new boolean[N];

        DFS(0);
        System.out.println(sb);
    }

    static void DFS(int depth){
        if(depth == M){
            for(int n : arr){
                sb.append(n).append(" ");
            }
            sb.append("\n");
            return;
        }

        for(int i=0;i<N;i++){
            if(!vis[i]){
                vis[i] = true;
                arr[depth] = i + 1;
                DFS(depth + 1);
                vis[i]=false;
            }
        }
    }
}
